src/messenger/messenger.cpp

```diff
@@ -398,128 +404,225 @@ void Messenger::Terminate() {
   is_messenger_running_ = false;
 }
 
-void Messenger::ListenForConnection(const ConnectionDestination &target, const std::string &identity,
-                                    CallbackFn callback) {
+router_id_t Messenger::ListenForConnection(const ConnectionDestination &target, const std::string &identity,
+                                           CallbackFn callback) {
   // ZeroMQ is not thread-safe, and the actual binding of new connection endpoints must be done from the same
   // thread that is going to poll the endpoints. See RouterToBeAdded docstring.
   std::unique_lock lock(routers_add_mutex_);
-  routers_to_be_added_.emplace_back(RouterToBeAdded{target, identity, std::move(callback)});
+  router_id_t router_id = next_router_id_++;
+  routers_to_be_added_.emplace_back(RouterToBeAdded{router_id, target, identity, std::move(callback)});
   routers_add_cvar_.wait(lock);
+  return router_id;
 }
 
-ConnectionId Messenger::MakeConnection(const ConnectionDestination &target) {
-  return ConnectionId(common::ManagedPointer(this), target, identity_);
+connection_id_t Messenger::MakeConnection(const ConnectionDestination &target) {
+  // ZeroMQ is not thread-safe, and the actual creation of new sockets must be done from the same
+  // thread that is going to send and receive using those sockets. See ConnectionToBeAdded docstring.
+  std::unique_lock lock(connections_add_mutex_);
+  connection_id_t connection_id = next_connection_id_++;
```

Comment:
I'm not sure if this matters, but since `next_connection_id_` is only accessed with a lock it might not need to be an atomic int.

Reply:
good point, will add to comment